src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +319,99 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeoutDelay q = do
+  timedOutVar ← registerDelay timeoutDelay
+
+  let
+    readItems xs =
+      readTBMQueue q ≫= \case
+        Nothing → return xs
+        Just x → go (x:xs)
+
+    timeout =
+      readTVar timedOutVar ≫= check
+
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = readItems xs <|> xs <$ timeout
+
+  atomically $ go []
```

Comment:
I think putting `go` into a single atomic transaction is not a good idea here (sorry for proposing it). It will wait until there are enough items in the queue or the timeout triggers, but it will do so in a very busy way, by counting the available items (starting from zero) each time a new item is added to the list. At least it won't interfer with the writers, since it's retrying most of the time, thus not causing any harm. However, since the final successful transaction is pretty large chances are the the writers interfer with the that successful read operation if it involves a role-over of the internal read and write lists.

I think the following code is better:

```
let
  readItems xs = atomically (readTBMQueue q) ≫= \case
    Nothing → return xs
    Just x → go (x:xs)

  go xs
    | length xs ≥ n = return xs
    | otherwise = do 
        t ← atomically $ readTVar timedOutVar
        if not t || null xs then readItems xs else return xs
```